File: sdks/sandbox/kotlin/sandbox/src/main/kotlin/com/alibaba/opensandbox/sandbox/mount/MountExtensions.kt
/*
 * Copyright 2025 Alibaba Group Holding Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.alibaba.opensandbox.sandbox.mount

import com.alibaba.opensandbox.sandbox.Sandbox
import com.alibaba.opensandbox.sandbox.domain.exceptions.InvalidArgumentException
import com.alibaba.opensandbox.sandbox.domain.exceptions.MountFailedException
import com.alibaba.opensandbox.sandbox.domain.models.execd.executions.Execution
import com.alibaba.opensandbox.sandbox.domain.models.execd.filesystem.WriteEntry
import com.alibaba.opensandbox.sandbox.domain.models.mount.NfsMountOptions
import com.alibaba.opensandbox.sandbox.domain.models.mount.OssfsMountOptions
import java.util.UUID

/**
 * Syntax sugar that mounts an NFS export inside the sandbox by running the
 * appropriate `mount -t nfs` command via [Sandbox.commands].
 *
 * The remote sandbox must already have `mount.nfs` available, or
 * [NfsMountOptions.installation] must install it.
 *
 * @throws InvalidArgumentException if the options are invalid.
 * @throws MountFailedException if the mount command fails (non-zero exit or `error` event).
 */
@JvmName("mountNfs")
fun Sandbox.mount(options: NfsMountOptions): Execution {
    validateNfs(options)
    val optString = options.options?.takeIf { it.isNotBlank() } ?: NfsMountOptions.DEFAULT_NFS_OPTIONS
    val core =
        "mkdir -p ${shQuote(options.mountPoint)} && " +
            "mount -t nfs -o ${shQuote(optString)} " +
            "${shQuote(options.endpoint + ":" + options.nasPath)} ${shQuote(options.mountPoint)}"
    val cmd = prependInstallation(options.installation, core)
    val execution = commands().run(cmd)
    ensureSuccess(execution, "NAS mount failure")
    return execution
}

/**
 * Syntax sugar that mounts an Alibaba Cloud OSS bucket inside the sandbox using
 * `ossfs` (1.x) or `ossfs2` (2.x), selected by [OssfsMountOptions.version].
 *
 * @throws InvalidArgumentException if the options are invalid.
 * @throws MountFailedException if the mount command fails.
 */
@JvmName("mountOssfs")
fun Sandbox.mount(options: OssfsMountOptions): Execution {
    validateOssfs(options)
    return when (options.version ?: OssfsMountOptions.Version.OSSFS_1_0) {
        OssfsMountOptions.Version.OSSFS_1_0 -> mountOssfs1(options)
        OssfsMountOptions.Version.OSSFS_2_0 -> mountOssfs2(options)
    }
}

/**
 * Unmount a previously mounted path inside the sandbox using `umount`.
 *
 * @throws MountFailedException if the umount command fails.
 */
fun Sandbox.umount(mountPoint: String): Execution {
    if (mountPoint.isBlank()) {
        throw InvalidArgumentException(message = "mountPoint must not be blank")
    }
    val execution = commands().run("umount ${shQuote(mountPoint)}")
    ensureSuccess(execution, "umount failure")
    return execution
}

private fun Sandbox.mountOssfs1(options: OssfsMountOptions): Execution {
    val useSts = !options.securityToken.isNullOrBlank()
    // ossfs 1.x password file format:
    //   AK/SK mode : bucket:accessKeyId:accessKeySecret
    //   STS mode   : bucket:accessKeyId:accessKeySecret:securityToken
    val passwd =
        buildString {
            append(options.bucket).append(':')
            append(options.accessKeyId).append(':')
            append(options.accessKeySecret)
            if (useSts) {
                append(':').append(options.securityToken)
            }
        }
    val bucketArg =
        if (!options.bucketDirectory.isNullOrBlank()) {
            "${options.bucket}:/${options.bucketDirectory}"
        } else {
            options.bucket
        }
    val optionFlags =
        options.options.joinToString(separator = "") { " -o${shQuote(it)}" }

    // Use a unique per-call password file under /tmp so that concurrent
    // ossfs 1.x mounts do not overwrite or delete each other's credentials.
    // Create it with mode 0600 in one atomic step (avoids the world-readable
    // window that a naive `echo > ; chmod 600` would create). Always clean it
    // up even if ossfs fails, by preserving the subshell rc.
    val passwdPath = "/tmp/opensandbox-ossfspass-${UUID.randomUUID()}"
    val quotedPasswdPath = shQuote(passwdPath)
    val core =
        "ossfs --version && " +
            "install -m 600 /dev/null $quotedPasswdPath && " +
            "printf %s ${shQuote(passwd)} > $quotedPasswdPath && " +
            "mkdir -p ${shQuote(options.mountPoint)} && " +
            "( ossfs ${shQuote(bucketArg)} ${shQuote(options.mountPoint)} " +
            "-ourl=${shQuote(options.endpoint)} " +
            "-opasswd_file=$quotedPasswdPath$optionFlags ); " +
            "__rc=\$?; rm -f $quotedPasswdPath; exit \$__rc"

    val cmd = prependInstallation(options.installation, core)
    val execution = commands().run(cmd)
    ensureSuccess(execution, "ossfs1.0 mount failure")
    return execution
}

private fun Sandbox.mountOssfs2(options: OssfsMountOptions): Execution {
    val confBuilder = StringBuilder()
    confBuilder.append("--oss_endpoint=").append(options.endpoint).append('\n')
    confBuilder.append("--oss_bucket=").append(options.bucket).append('\n')
    if (!options.bucketDirectory.isNullOrBlank()) {
        // ossfs2 mounts a bucket root; a subdirectory is expressed as a prefix
        // (trailing slash makes it a directory boundary).
        val prefix = options.bucketDirectory.trimEnd('/') + "/"
        confBuilder.append("--oss_bucket_prefix=").append(prefix).append('\n')
    }
    for (opt in options.options) {
        confBuilder.append("--").append(opt).append('\n')
    }

    val confPath = "/tmp/opensandbox-ossfs-${UUID.randomUUID()}.conf"
    files().writeFile(
        WriteEntry.builder()
            .path(confPath)
            .data(confBuilder.toString())
            // Mode is serialized as a JSON number and parsed by execd as an
            // octal string (see components/execd/pkg/web/controller/utils.go);
            // pass the literal decimal 600 to get filesystem mode 0o600.
            .mode(600)
            .build(),
    )

    // ossfs 2.x reads credentials from environment variables:
    //   AK/SK mode : OSS_ACCESS_KEY_ID + OSS_ACCESS_KEY_SECRET
    //   STS mode   : additionally OSS_SESSION_TOKEN
    val stsExport =
        if (!options.securityToken.isNullOrBlank()) {
            " && export OSS_SESSION_TOKEN=${shQuote(options.securityToken)}"
        } else {
            ""
        }

    val core =
        "ossfs2 --version && " +
            "mkdir -p ${shQuote(options.mountPoint)} && " +
            "export OSS_ACCESS_KEY_ID=${shQuote(options.accessKeyId)} && " +
            "export OSS_ACCESS_KEY_SECRET=${shQuote(options.accessKeySecret)}" +
            stsExport +
            " && ossfs2 mount ${shQuote(options.mountPoint)} -c ${shQuote(confPath)}"

    val cmd = prependInstallation(options.installation, core)
    val execution = commands().run(cmd)
    ensureSuccess(execution, "ossfs2.0 mount failure")
    return execution
}

private fun validateNfs(options: NfsMountOptions) {
    if (options.endpoint.isBlank()) throw InvalidArgumentException(message = "endpoint must not be blank")
    if (options.mountPoint.isBlank()) throw InvalidArgumentException(message = "mountPoint must not be blank")
    if (options.nasPath.isBlank()) throw InvalidArgumentException(message = "nasPath must not be blank")
}

private fun validateOssfs(options: OssfsMountOptions) {
    if (options.endpoint.isBlank()) throw InvalidArgumentException(message = "endpoint must not be blank")
    if (options.bucket.isBlank()) throw InvalidArgumentException(message = "bucket must not be blank")
    if (options.mountPoint.isBlank()) throw InvalidArgumentException(message = "mountPoint must not be blank")
    if (options.accessKeyId.isBlank()) throw InvalidArgumentException(message = "accessKeyId must not be blank")
    if (options.accessKeySecret.isBlank()) throw InvalidArgumentException(message = "accessKeySecret must not be blank")
}

private fun ensureSuccess(
    execution: Execution,
    failurePrefix: String,
) {
    val error = execution.error
    val exitCode = execution.exitCode
    val failed = error != null || (exitCode != null && exitCode != 0)
    if (!failed) return
    val stderr = execution.logs.stderr.joinToString("\n") { it.text }
    val detail =
        buildString {
            if (error != null) {
                append('[').append(error.name).append("] ").append(error.value)
            }
            if (stderr.isNotEmpty()) {
                if (isNotEmpty()) append(" | ")
                append("stderr=").append(stderr)
            }
        }
    throw MountFailedException(
        message = if (detail.isNotEmpty()) "$failurePrefix: $detail" else failurePrefix,
        execution = execution,
    )
}

private fun prependInstallation(
    installation: String?,
    core: String,
): String {
    return if (!installation.isNullOrBlank()) "$installation && $core" else core
}

/**
 * Quote a value for POSIX shell single-quoted context.
 *
 * `'` inside the value is escaped as `'\''`. The result is always safe to embed
 * as one argument to a `sh -c` string.
 */
internal fun shQuote(value: String): String {
    val escaped = value.replace("'", "'\\''")
    return "'$escaped'"
}

File: sdks/sandbox/csharp/src/OpenSandbox/MountExtensions.cs
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using OpenSandbox.Core;
using OpenSandbox.Internal;
using OpenSandbox.Models;

namespace OpenSandbox;

/// <summary>
/// Extension methods that add NAS/OSS mount syntax sugar to <see cref="Sandbox"/>.
/// </summary>
/// <remarks>
/// These helpers do not manage the remote NAS/OSS service. They assemble the
/// appropriate <c>mount -t nfs</c> / <c>ossfs</c> / <c>ossfs2</c> shell command
/// and run it via <see cref="Sandbox.Commands"/>. The sandbox image must have
/// the corresponding mount binary installed, or <c>Installation</c> can be used
/// to install it at mount time.
/// </remarks>
public static class MountExtensions
{
    /// <summary>
    /// Mount a NAS export inside the sandbox by running <c>mount -t nfs</c>.
    /// </summary>
    /// <param name="sandbox">The sandbox to mount into.</param>
    /// <param name="options">The NFS mount options.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The <see cref="Execution"/> from the underlying <c>RunAsync</c> call.</returns>
    /// <exception cref="InvalidArgumentException">Thrown when <paramref name="options"/> is invalid.</exception>
    /// <exception cref="MountFailedException">Thrown when the mount command fails inside the sandbox.</exception>
    public static async Task<Execution> MountAsync(
        this Sandbox sandbox,
        NfsMountOptions options,
        CancellationToken cancellationToken = default)
    {
        if (sandbox is null)
        {
            throw new ArgumentNullException(nameof(sandbox));
        }
        if (options is null)
        {
            throw new ArgumentNullException(nameof(options));
        }
        MountShell.ValidateNfs(options);
        var execution = await sandbox.Commands
            .RunAsync(MountShell.BuildNfsCommand(options), cancellationToken: cancellationToken)
            .ConfigureAwait(false);
        MountShell.EnsureSuccess(execution, "NAS mount failure");
        return execution;
    }

    /// <summary>
    /// Mount an Alibaba Cloud OSS bucket inside the sandbox using either
    /// <c>ossfs</c> (1.x) or <c>ossfs2</c> (2.x), selected by
    /// <see cref="OssfsMountOptions.Version"/>.
    /// </summary>
    /// <param name="sandbox">The sandbox to mount into.</param>
    /// <param name="options">The OSS mount options.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The <see cref="Execution"/> from the underlying <c>RunAsync</c> call.</returns>
    /// <exception cref="InvalidArgumentException">Thrown when <paramref name="options"/> is invalid.</exception>
    /// <exception cref="MountFailedException">Thrown when the mount command fails inside the sandbox.</exception>
    public static async Task<Execution> MountAsync(
        this Sandbox sandbox,
        OssfsMountOptions options,
        CancellationToken cancellationToken = default)
    {
        if (sandbox is null)
        {
            throw new ArgumentNullException(nameof(sandbox));
        }
        if (options is null)
        {
            throw new ArgumentNullException(nameof(options));
        }
        MountShell.ValidateOssfs(options);
        var version = MountShell.SelectOssfsVersion(options);
        if (version == OssfsVersion.Ossfs10)
        {
            var execution = await sandbox.Commands
                .RunAsync(MountShell.BuildOssfs1Command(options), cancellationToken: cancellationToken)
                .ConfigureAwait(false);
            MountShell.EnsureSuccess(execution, "ossfs1.0 mount failure");
            return execution;
        }
        var plan = MountShell.BuildOssfs2Plan(options);
        await sandbox.Files
            .WriteFilesAsync(new[] { MountShell.BuildOssfs2ConfEntry(plan) }, cancellationToken)
            .ConfigureAwait(false);
        var exec = await sandbox.Commands
            .RunAsync(plan.Command, cancellationToken: cancellationToken)
            .ConfigureAwait(false);
        MountShell.EnsureSuccess(exec, "ossfs2.0 mount failure");
        return exec;
    }

    /// <summary>
    /// Unmount a previously mounted path inside the sandbox by running <c>umount</c>.
    /// </summary>
    /// <param name="sandbox">The sandbox to unmount from.</param>
    /// <param name="mountPoint">The absolute path inside the sandbox to unmount.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The <see cref="Execution"/> from the underlying <c>RunAsync</c> call.</returns>
    /// <exception cref="InvalidArgumentException">Thrown when <paramref name="mountPoint"/> is blank.</exception>
    /// <exception cref="MountFailedException">Thrown when the umount command fails.</exception>
    public static async Task<Execution> UmountAsync(
        this Sandbox sandbox,
        string mountPoint,
        CancellationToken cancellationToken = default)
    {
        if (sandbox is null)
        {
            throw new ArgumentNullException(nameof(sandbox));
        }
        var command = MountShell.BuildUmountCommand(mountPoint);
        var execution = await sandbox.Commands
            .RunAsync(command, cancellationToken: cancellationToken)
            .ConfigureAwait(false);
        MountShell.EnsureSuccess(execution, "umount failure");
        return execution;
    }
}
